### PR TITLE
[Tests] More resilient network tests, take 2

### PR DESCRIPTION
**Pull request by grendello**
The previous
commit (https://github.com/xamarin/xamarin-android/commit/608d950a703ddb527258565c2a0705aa37f71789)
solved only part of the issue with network tests which fail on connection to a
test server. Not only one test was left without any code to catch failed
connection attempts, all the other tests failed to catch a connection issue in
situations when the Java backend tries to connect to the server only after the
managed code attempts to wait on or obtain the result of the request from the
asynchronous task running the request.

This commit *should* solve the remaining problems.

### CODE REVIEW

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs [hunk 1]
```diff
@@ -151,15 +151,37 @@ protected Task<HttpResponseMessage> ConnectIgnoreFailure (Func<Task<HttpResponse
...
+			try {
+				runner ();
+			} catch (AggregateException ex) {
+				if (IgnoreIfConnectionFailed (ex.InnerException as WebException, out connectionFailed))
+					return;
 				throw;
 			}
 		}
+
+		bool IgnoreIfConnectionFailed (WebException wex, out bool connectionFailed)
```
Comment (jonpryor) [inline src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: Why does this need an `out bool` parameter, and/or why does this need a `bool` return value?  As far as I can tell, every codepath has `connectionFailed==return-value`.

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs [hunk 2]
```diff
@@ -151,15 +151,37 @@ protected Task<HttpResponseMessage> ConnectIgnoreFailure (Func<Task<HttpResponse
 			try {
 				return connector ();
 			} catch (AggregateException ex) {
-				if (ex.InnerException is WebException wex && wex.Status == WebExceptionStatus.ConnectFailure) {
-					connectionFailed = true;
-					Assert.Ignore ($"Failed to connect to server. {wex}");
+				if (IgnoreIfConnectionFailed (ex.InnerException as WebException, out connectionFailed))
```
Comment (jonpryor) [inline src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: Couldn't there be any number of inner exceptions that the `AggregateException` wraps?  Therefore, wouldn't it be prudent to examine *both* `AggregateException.InnerExceptions` *and* the `InnerException` chain?

```csharp
static IEnumerable<Exception> InnerExceptions (Exception e)
{
    while (e.InnerException != null) {
        yield return e.InnerException;
        e = e.InnerException;
    }
}
...
if (ex.InnerExceptions.SelectMany (ie => InnerExceptions (ie))
        .Any (ie => IgnoreIfConnectionFailed (ie as WebException, out connectionFailed))) { 
    return null;
}
```

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs [hunk 3]
```diff
@@ -151,15 +151,37 @@ protected Task<HttpResponseMessage> ConnectIgnoreFailure (Func<Task<HttpResponse
...
+			}
+		}
 
+		protected void RunIgnoringNetworkIssues (Action runner, out bool connectionFailed)
+		{
+			connectionFailed = false;
+			try {
+				runner ();
+			} catch (AggregateException ex) {
+				if (IgnoreIfConnectionFailed (ex.InnerException as WebException, out connectionFailed))
```
Comment (jonpryor) [inline src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: ...ditto here?

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs (hunk 2 above)
Comment (grendello) [reply src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: We don't want to catch any exception somewhere in the chain - we know that "our" exception we want to ignore is always the AggregateException's `InnerException`.

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs (hunk 1 above)
Comment (grendello) [reply src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: The `out bool` is to stay DRY - this function is called from two different methods which have `out bool conectionFailed` used by **their** caller. And the return `bool` is to indicate whether to ignore the exception thrown or rethrow it.

File: src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs (hunk 3 above)
Comment (grendello) [reply src/Mono.Android/Test/Xamarin.Android.Net/AndroidClientHandlerTests.cs]: Ditto as well

**Pull request merged**